#***Carlos Gross-Martinez***
#***CAP 6635 - Artificial Intelligence***

#***Question 7a***

In [ ]:
# Min_Max Decision playing Tic-Tac-Toe
# Code adapted from: https://stackabuse.com/minimax-and-alpha-beta-pruning-in-python/

import time

class Game:
    def __init__(self):
        self.initialize_game()

    def initialize_game(self):
        self.current_state = [['.','.','.'],
                              ['.','.','.'],
                              ['.','.','.']]

        # Player X always plays first
        self.player_turn = 'X'

    def draw_board(self):
        for i in range(0, 3):
            for j in range(0, 3):
                print('{}|'.format(self.current_state[i][j]), end=" ")
            print()
        print()

    # Determines if the made move is a legal move
    def is_valid(self, px, py):
        if px < 0 or px > 2 or py < 0 or py > 2:
            return False
        elif self.current_state[px][py] != '.':
            return False
        else:
            return True

    # Checks if the game has ended and returns the winner in each case
    def is_end(self):
        # Vertical win
        for i in range(0, 3):
            if (self.current_state[0][i] != '.' and
                self.current_state[0][i] == self.current_state[1][i] and
                self.current_state[1][i] == self.current_state[2][i]):
                return self.current_state[0][i]

        # Horizontal win
        for i in range(0, 3):
            if (self.current_state[i] == ['X', 'X', 'X']):
                return 'X'
            elif (self.current_state[i] == ['O', 'O', 'O']):
                return 'O'

        # Main diagonal win
        if (self.current_state[0][0] != '.' and
            self.current_state[0][0] == self.current_state[1][1] and
            self.current_state[0][0] == self.current_state[2][2]):
            return self.current_state[0][0]

        # Second diagonal win
        if (self.current_state[0][2] != '.' and
            self.current_state[0][2] == self.current_state[1][1] and
            self.current_state[0][2] == self.current_state[2][0]):
            return self.current_state[0][2]

        # Is whole board full?
        for i in range(0, 3):
            for j in range(0, 3):
                # There's an empty field, we continue the game
                if (self.current_state[i][j] == '.'):
                    return None

        # It's a tie!
        return '.'

    # Player 'O' is max, in this case AI
    def max(self):

        # Possible values for maxv are:
        # -1 - loss
        # 0  - a tie
        # 1  - win

        # We're initially setting it to -2 as worse than the worst case:
        maxv = -2

        px = None
        py = None

        result = self.is_end()

        # If the game came to an end, the function needs to return
        # the evaluation function of the end. That can be:
        # -1 - loss
        # 0  - a tie
        # 1  - win
        if result == 'X':
            return (-1, 0, 0)
        elif result == 'O':
            return (1, 0, 0)
        elif result == '.':
            return (0, 0, 0)

        for i in range(0, 3):
            for j in range(0, 3):
                if self.current_state[i][j] == '.':
                    # On the empty field player 'O' makes a move and calls Min
                    # That's one branch of the game tree.
                    self.current_state[i][j] = 'O'
                    (m, min_i, min_j) = self.min()
                    # Fixing the maxv value if needed
                    if m > maxv:
                        maxv = m
                        px = i
                        py = j
                    # Setting back the field to empty
                    self.current_state[i][j] = '.'
        return (maxv, px, py)

    # Player 'X' is min, in this case human
    def min(self):

    # Possible values for minv are:
    # -1 - win
    # 0  - a tie
    # 1  - loss

    # We're initially setting it to 2 as worse than the worst case:
        minv = 2

        qx = None
        qy = None

        result = self.is_end()

        if result == 'X':
            return (-1, 0, 0)
        elif result == 'O':
            return (1, 0, 0)
        elif result == '.':
            return (0, 0, 0)

        for i in range(0, 3):
            for j in range(0, 3):
                if self.current_state[i][j] == '.':
                    self.current_state[i][j] = 'X'
                    (m, max_i, max_j) = self.max()
                    if m < minv:
                        minv = m
                        qx = i
                        qy = j
                    self.current_state[i][j] = '.'

        return (minv, qx, qy)

    def play(self):
        totalTime=0
        while True:
            self.draw_board()
            self.result = self.is_end()

            # Printing the appropriate message if the game has ended
            if self.result != None:
                if self.result == 'X':
                    print('The winner is X!')
                elif self.result == 'O':
                    print('The winner is O!')
                elif self.result == '.':
                    print("It's a tie!")

                self.initialize_game()
                print('The total Evaluation time:{}s'.format(round(totalTime, 7)))
                return

            # If it's player's turn
            if self.player_turn == 'X':

                while True:

                    start = time.time()
                    (m, qx, qy) = self.min()
                    end = time.time()
                    totalTime=totalTime+(end-start)
                    print('Evaluation time: {}s'.format(round(end - start, 7)))
                    print('Recommended move: X = {}, Y = {}'.format(qx, qy))

                    px = int(input('Insert the X coordinate: '))
                    py = int(input('Insert the Y coordinate: '))

                    (qx, qy) = (px, py)

                    if self.is_valid(px, py):
                        self.current_state[px][py] = 'X'
                        self.player_turn = 'O'
                        break
                    else:
                        print('The move is not valid! Try again.')

            # If it's AI's turn
            else:
                (m, px, py) = self.max()
                self.current_state[px][py] = 'O'
                self.player_turn = 'X'

In [ ]:
#Game 1
def main():
    g = Game()
    g.play()

if __name__ == "__main__":
    main()

.| .| .| 
.| .| .| 
.| .| .| 

Evaluation time: 2.0595455s
Recommended move: X = 0, Y = 0
Insert the X coordinate: 0
Insert the Y coordinate: 0
X| .| .| 
.| .| .| 
.| .| .| 

X| .| .| 
.| O| .| 
.| .| .| 

Evaluation time: 0.0268316s
Recommended move: X = 0, Y = 1
Insert the X coordinate: 2
Insert the Y coordinate: 2
X| .| .| 
.| O| .| 
.| .| X| 

X| O| .| 
.| O| .| 
.| .| X| 

Evaluation time: 0.0008419s
Recommended move: X = 2, Y = 1
Insert the X coordinate: 2
Insert the Y coordinate: 1
X| O| .| 
.| O| .| 
.| X| X| 

X| O| .| 
.| O| .| 
O| X| X| 

Evaluation time: 8.2e-05s
Recommended move: X = 0, Y = 2
Insert the X coordinate: 0
Insert the Y coordinate: 2
X| O| X| 
.| O| .| 
O| X| X| 

X| O| X| 
.| O| O| 
O| X| X| 

Evaluation time: 1.6e-05s
Recommended move: X = 1, Y = 0
Insert the X coordinate: 1
Insert the Y coordinate: 0
X| O| X| 
X| O| O| 
O| X| X| 

It's a tie!
The total Evaluation time:2.087317s


In [ ]:
#Game 2
def main():
    g = Game()
    g.play()

if __name__ == "__main__":
    main()

.| .| .| 
.| .| .| 
.| .| .| 

Evaluation time: 1.9209108s
Recommended move: X = 0, Y = 0
Insert the X coordinate: 0
Insert the Y coordinate: 0
X| .| .| 
.| .| .| 
.| .| .| 

X| .| .| 
.| O| .| 
.| .| .| 

Evaluation time: 0.0269904s
Recommended move: X = 0, Y = 1
Insert the X coordinate: 0
Insert the Y coordinate: 1
X| X| .| 
.| O| .| 
.| .| .| 

X| X| O| 
.| O| .| 
.| .| .| 

Evaluation time: 0.0006504s
Recommended move: X = 2, Y = 0
Insert the X coordinate: 2
Insert the Y coordinate: 0
X| X| O| 
.| O| .| 
X| .| .| 

X| X| O| 
O| O| .| 
X| .| .| 

Evaluation time: 8.15e-05s
Recommended move: X = 1, Y = 2
Insert the X coordinate: 1
Insert the Y coordinate: 2
X| X| O| 
O| O| X| 
X| .| .| 

X| X| O| 
O| O| X| 
X| O| .| 

Evaluation time: 1.65e-05s
Recommended move: X = 2, Y = 2
Insert the X coordinate: 2
Insert the Y coordinate: 2
X| X| O| 
O| O| X| 
X| O| X| 

It's a tie!
The total Evaluation time:1.9486496s


In [ ]:
#Game 3
def main():
    g = Game()
    g.play()

if __name__ == "__main__":
    main()

.| .| .| 
.| .| .| 
.| .| .| 

Evaluation time: 1.6551619s
Recommended move: X = 0, Y = 0
Insert the X coordinate: 0
Insert the Y coordinate: 0
X| .| .| 
.| .| .| 
.| .| .| 

X| .| .| 
.| O| .| 
.| .| .| 

Evaluation time: 0.0236092s
Recommended move: X = 0, Y = 1
Insert the X coordinate: 2
Insert the Y coordinate: 2
X| .| .| 
.| O| .| 
.| .| X| 

X| O| .| 
.| O| .| 
.| .| X| 

Evaluation time: 0.0007994s
Recommended move: X = 2, Y = 1
Insert the X coordinate: 2
Insert the Y coordinate: 1
X| O| .| 
.| O| .| 
.| X| X| 

X| O| .| 
.| O| .| 
O| X| X| 

Evaluation time: 7.46e-05s
Recommended move: X = 0, Y = 2
Insert the X coordinate: 0
Insert the Y coordinate: 2
X| O| X| 
.| O| .| 
O| X| X| 

X| O| X| 
.| O| O| 
O| X| X| 

Evaluation time: 1.65e-05s
Recommended move: X = 1, Y = 0
Insert the X coordinate: 1
Insert the Y coordinate: 0
X| O| X| 
X| O| O| 
O| X| X| 

It's a tie!
The total Evaluation time:1.6796615s


In [ ]:
#Game 4
def main():
    g = Game()
    g.play()

if __name__ == "__main__":
    main()

.| .| .| 
.| .| .| 
.| .| .| 

Evaluation time: 1.6512461s
Recommended move: X = 0, Y = 0
Insert the X coordinate: 2
Insert the Y coordinate: 2
.| .| .| 
.| .| .| 
.| .| X| 

.| .| .| 
.| O| .| 
.| .| X| 

Evaluation time: 0.0204926s
Recommended move: X = 0, Y = 0
Insert the X coordinate: 0
Insert the Y coordinate: 0
X| .| .| 
.| O| .| 
.| .| X| 

X| O| .| 
.| O| .| 
.| .| X| 

Evaluation time: 0.0025578s
Recommended move: X = 2, Y = 1
Insert the X coordinate: 2
Insert the Y coordinate: 1
X| O| .| 
.| O| .| 
.| X| X| 

X| O| .| 
.| O| .| 
O| X| X| 

Evaluation time: 0.0001128s
Recommended move: X = 0, Y = 2
Insert the X coordinate: 0
Insert the Y coordinate: 2
X| O| X| 
.| O| .| 
O| X| X| 

X| O| X| 
.| O| O| 
O| X| X| 

Evaluation time: 1.5e-05s
Recommended move: X = 1, Y = 0
Insert the X coordinate: 1
Insert the Y coordinate: 0
X| O| X| 
X| O| O| 
O| X| X| 

It's a tie!
The total Evaluation time:1.6744242s


In [ ]:
#Game 5
def main():
    g = Game()
    g.play()

if __name__ == "__main__":
    main()

.| .| .| 
.| .| .| 
.| .| .| 

Evaluation time: 1.7176342s
Recommended move: X = 0, Y = 0
Insert the X coordinate: 1
Insert the Y coordinate: 1
.| .| .| 
.| X| .| 
.| .| .| 

O| .| .| 
.| X| .| 
.| .| .| 

Evaluation time: 0.0202878s
Recommended move: X = 0, Y = 1
Insert the X coordinate: 0
Insert the Y coordinate: 1
O| X| .| 
.| X| .| 
.| .| .| 

O| X| .| 
.| X| .| 
.| O| .| 

Evaluation time: 0.0014224s
Recommended move: X = 1, Y = 0
Insert the X coordinate: 1
Insert the Y coordinate: 0
O| X| .| 
X| X| .| 
.| O| .| 

O| X| .| 
X| X| O| 
.| O| .| 

Evaluation time: 0.0001023s
Recommended move: X = 0, Y = 2
Insert the X coordinate: 0
Insert the Y coordinate: 2
O| X| X| 
X| X| O| 
.| O| .| 

O| X| X| 
X| X| O| 
O| O| .| 

Evaluation time: 1.57e-05s
Recommended move: X = 2, Y = 2
Insert the X coordinate: 2
Insert the Y coordinate: 2
O| X| X| 
X| X| O| 
O| O| X| 

It's a tie!
The total Evaluation time:1.7394624s


In [ ]:
#Game 6
def main():
    g = Game()
    g.play()

if __name__ == "__main__":
    main()

.| .| .| 
.| .| .| 
.| .| .| 

Evaluation time: 1.8623915s
Recommended move: X = 0, Y = 0
Insert the X coordinate: 0
Insert the Y coordinate: 0
X| .| .| 
.| .| .| 
.| .| .| 

X| .| .| 
.| O| .| 
.| .| .| 

Evaluation time: 0.0372508s
Recommended move: X = 0, Y = 1
Insert the X coordinate: 1
Insert the Y coordinate: 0
X| .| .| 
X| O| .| 
.| .| .| 

X| .| .| 
X| O| .| 
O| .| .| 

Evaluation time: 0.0008955s
Recommended move: X = 0, Y = 2
Insert the X coordinate: 0
Insert the Y coordinate: 2
X| .| X| 
X| O| .| 
O| .| .| 

X| O| X| 
X| O| .| 
O| .| .| 

Evaluation time: 8.42e-05s
Recommended move: X = 2, Y = 1
Insert the X coordinate: 2
Insert the Y coordinate: 1
X| O| X| 
X| O| .| 
O| X| .| 

X| O| X| 
X| O| O| 
O| X| .| 

Evaluation time: 1.84e-05s
Recommended move: X = 2, Y = 2
Insert the X coordinate: 2
Insert the Y coordinate: 2
X| O| X| 
X| O| O| 
O| X| X| 

It's a tie!
The total Evaluation time:1.9006402s


In [ ]:
#Game 7
def main():
    g = Game()
    g.play()

if __name__ == "__main__":
    main()

.| .| .| 
.| .| .| 
.| .| .| 

Evaluation time: 2.2150002s
Recommended move: X = 0, Y = 0
Insert the X coordinate: 0
Insert the Y coordinate: 0
X| .| .| 
.| .| .| 
.| .| .| 

X| .| .| 
.| O| .| 
.| .| .| 

Evaluation time: 0.0285664s
Recommended move: X = 0, Y = 1
Insert the X coordinate: 0
Insert the Y coordinate: 1
X| X| .| 
.| O| .| 
.| .| .| 

X| X| O| 
.| O| .| 
.| .| .| 

Evaluation time: 0.0012527s
Recommended move: X = 2, Y = 0
Insert the X coordinate: 2
Insert the Y coordinate: 0
X| X| O| 
.| O| .| 
X| .| .| 

X| X| O| 
O| O| .| 
X| .| .| 

Evaluation time: 8.37e-05s
Recommended move: X = 1, Y = 2
Insert the X coordinate: 1
Insert the Y coordinate: 2
X| X| O| 
O| O| X| 
X| .| .| 

X| X| O| 
O| O| X| 
X| O| .| 

Evaluation time: 1.48e-05s
Recommended move: X = 2, Y = 2
Insert the X coordinate: 2
Insert the Y coordinate: 2
X| X| O| 
O| O| X| 
X| O| X| 

It's a tie!
The total Evaluation time:2.2449176s


In [ ]:
#Game 8
def main():
    g = Game()
    g.play()

if __name__ == "__main__":
    main()


.| .| .| 
.| .| .| 
.| .| .| 

Evaluation time: 1.625169s
Recommended move: X = 0, Y = 0
Insert the X coordinate: 0
Insert the Y coordinate: 2
.| .| X| 
.| .| .| 
.| .| .| 

.| .| X| 
.| O| .| 
.| .| .| 

Evaluation time: 0.0219271s
Recommended move: X = 0, Y = 0
Insert the X coordinate: 0
Insert the Y coordinate: 0
X| .| X| 
.| O| .| 
.| .| .| 

X| O| X| 
.| O| .| 
.| .| .| 

Evaluation time: 0.0010321s
Recommended move: X = 2, Y = 1
Insert the X coordinate: 2
Insert the Y coordinate: 1
X| O| X| 
.| O| .| 
.| X| .| 

X| O| X| 
O| O| .| 
.| X| .| 

Evaluation time: 7.22e-05s
Recommended move: X = 1, Y = 2
Insert the X coordinate: 1
Insert the Y coordinate: 2
X| O| X| 
O| O| X| 
.| X| .| 

X| O| X| 
O| O| X| 
.| X| O| 

Evaluation time: 1.69e-05s
Recommended move: X = 2, Y = 0
Insert the X coordinate: 2
Insert the Y coordinate: 0
X| O| X| 
O| O| X| 
X| X| O| 

It's a tie!
The total Evaluation time:1.6482174s


In [ ]:
#Game 9
def main():
    g = Game()
    g.play()

if __name__ == "__main__":
    main()

.| .| .| 
.| .| .| 
.| .| .| 

Evaluation time: 1.648129s
Recommended move: X = 0, Y = 0
Insert the X coordinate: 0
Insert the Y coordinate: 2
.| .| X| 
.| .| .| 
.| .| .| 

.| .| X| 
.| O| .| 
.| .| .| 

Evaluation time: 0.0211809s
Recommended move: X = 0, Y = 0
Insert the X coordinate: 0
Insert the Y coordinate: 1
.| X| X| 
.| O| .| 
.| .| .| 

O| X| X| 
.| O| .| 
.| .| .| 

Evaluation time: 0.0010161s
Recommended move: X = 2, Y = 2
Insert the X coordinate: 2
Insert the Y coordinate: 2
O| X| X| 
.| O| .| 
.| .| X| 

O| X| X| 
.| O| O| 
.| .| X| 

Evaluation time: 8.85e-05s
Recommended move: X = 1, Y = 0
Insert the X coordinate: 1
Insert the Y coordinate: 0
O| X| X| 
X| O| O| 
.| .| X| 

O| X| X| 
X| O| O| 
O| .| X| 

Evaluation time: 1.79e-05s
Recommended move: X = 2, Y = 1
Insert the X coordinate: 2
Insert the Y coordinate: 1
O| X| X| 
X| O| O| 
O| X| X| 

It's a tie!
The total Evaluation time:1.6704323s


In [ ]:
#Game 10
def main():
    g = Game()
    g.play()

if __name__ == "__main__":
    main()

.| .| .| 
.| .| .| 
.| .| .| 

Evaluation time: 2.419281s
Recommended move: X = 0, Y = 0
Insert the X coordinate: 0
Insert the Y coordinate: 0
X| .| .| 
.| .| .| 
.| .| .| 

X| .| .| 
.| O| .| 
.| .| .| 

Evaluation time: 0.0247066s
Recommended move: X = 0, Y = 1
Insert the X coordinate: 2
Insert the Y coordinate: 1
X| .| .| 
.| O| .| 
.| X| .| 

X| .| .| 
O| O| .| 
.| X| .| 

Evaluation time: 0.0009706s
Recommended move: X = 1, Y = 2
Insert the X coordinate: 1
Insert the Y coordinate: 2
X| .| .| 
O| O| X| 
.| X| .| 

X| .| O| 
O| O| X| 
.| X| .| 

Evaluation time: 7.51e-05s
Recommended move: X = 2, Y = 0
Insert the X coordinate: 2
Insert the Y coordinate: 0
X| .| O| 
O| O| X| 
X| X| .| 

X| .| O| 
O| O| X| 
X| X| O| 

Evaluation time: 1.6e-05s
Recommended move: X = 0, Y = 1
Insert the X coordinate: 0
Insert the Y coordinate: 1
X| X| O| 
O| O| X| 
X| X| O| 

It's a tie!
The total Evaluation time:2.4450493s


#Avg Comp Thik Time = (2.087 + 1.947 + 1.680 + 1.674 + 1.739 + 1.901 + 2.225 + 1.648 + 1.670 + 2.445) / 10

#Avg Comp Thik Time = 1.902

#It is hard to win agains the computer becuase it will always search for the state which will allow it to win or tie the game if the player makes a mistake. It is able to create a tree with all posible game plays and therefore has a insight on what is the next best move based on the current state which will allow it to win.

#***Question 7b***

In [ ]:
# CAP 6635 Artificial Intelligence
# Alpha_Beta Pruning playing Tic-Tac-Toe
# Code adapted from: https://stackabuse.com/minimax-and-alpha-beta-pruning-in-python/

import time

class Game:
    def __init__(self):
        self.initialize_game()

    def initialize_game(self):
        self.current_state = [['.','.','.'],
                              ['.','.','.'],
                              ['.','.','.']]

        # Player X always plays first
        self.player_turn = 'X'

    def draw_board(self):
        for i in range(0, 3):
            for j in range(0, 3):
                print('{}|'.format(self.current_state[i][j]), end=" ")
            print()
        print()

    # Determines if the made move is a legal move
    def is_valid(self, px, py):
        if px < 0 or px > 2 or py < 0 or py > 2:
            return False
        elif self.current_state[px][py] != '.':
            return False
        else:
            return True

    # Checks if the game has ended and returns the winner in each case
    def is_end(self):
        # Vertical win
        for i in range(0, 3):
            if (self.current_state[0][i] != '.' and
                self.current_state[0][i] == self.current_state[1][i] and
                self.current_state[1][i] == self.current_state[2][i]):
                return self.current_state[0][i]

        # Horizontal win
        for i in range(0, 3):
            if (self.current_state[i] == ['X', 'X', 'X']):
                return 'X'
            elif (self.current_state[i] == ['O', 'O', 'O']):
                return 'O'

        # Main diagonal win
        if (self.current_state[0][0] != '.' and
            self.current_state[0][0] == self.current_state[1][1] and
            self.current_state[0][0] == self.current_state[2][2]):
            return self.current_state[0][0]

        # Second diagonal win
        if (self.current_state[0][2] != '.' and
            self.current_state[0][2] == self.current_state[1][1] and
            self.current_state[0][2] == self.current_state[2][0]):
            return self.current_state[0][2]

        # Is whole board full?
        for i in range(0, 3):
            for j in range(0, 3):
                # There's an empty field, we continue the game
                if (self.current_state[i][j] == '.'):
                    return None

        # It's a tie!
        return '.'

    # Player 'O' is max, in this case AI
    def max_alpha_beta(self, alpha, beta):
        maxv = -2
        px = None
        py = None

        result = self.is_end()

        if result == 'X':
            return (-1, 0, 0)
        elif result == 'O':
            return (1, 0, 0)
        elif result == '.':
            return (0, 0, 0)

        for i in range(0, 3):
            for j in range(0, 3):
                if self.current_state[i][j] == '.':
                    self.current_state[i][j] = 'O'
                    (m, min_i, in_j) = self.min_alpha_beta(alpha, beta)
                    if m > maxv:
                        maxv = m
                        px = i
                        py = j
                    self.current_state[i][j] = '.'

                    # Next two ifs in Max and Min are the only difference between regular algorithm and minimax
                    if maxv >= beta:
                        return (maxv, px, py)

                    if maxv > alpha:
                        alpha = maxv

        return (maxv, px, py)

    # Player 'X' is min, in this case human
    def min_alpha_beta(self, alpha, beta):

        minv = 2

        qx = None
        qy = None

        result = self.is_end()

        if result == 'X':
            return (-1, 0, 0)
        elif result == 'O':
            return (1, 0, 0)
        elif result == '.':
            return (0, 0, 0)

        for i in range(0, 3):
            for j in range(0, 3):
                if self.current_state[i][j] == '.':
                    self.current_state[i][j] = 'X'
                    (m, max_i, max_j) = self.max_alpha_beta(alpha, beta)
                    if m < minv:
                        minv = m
                        qx = i
                        qy = j
                    self.current_state[i][j] = '.'

                    if minv <= alpha:
                        return (minv, qx, qy)

                    if minv < beta:
                        beta = minv

        return (minv, qx, qy)


    def play_alpha_beta(self):
        totalTime=0
        while True:
            self.draw_board()
            self.result = self.is_end()

            if self.result != None:
                if self.result == 'X':
                    print('The winner is X!')
                elif self.result == 'O':
                    print('The winner is O!')
                elif self.result == '.':
                    print("It's a tie!")


                self.initialize_game()
                print('The total Evaluation time:{}s'.format(round(totalTime, 7)))
                return

            if self.player_turn == 'X':

                while True:
                    start = time.time()
                    (m, qx, qy) = self.min_alpha_beta(-2, 2)
                    end = time.time()
                    totalTime=totalTime+(end-start)
                    print('Evaluation time: {}s'.format(round(end - start, 7)))
                    print('Recommended move: X = {}, Y = {}'.format(qx, qy))

                    px = int(input('Insert the X coordinate: '))
                    py = int(input('Insert the Y coordinate: '))

                    qx = px
                    qy = py

                    if self.is_valid(px, py):
                        self.current_state[px][py] = 'X'
                        self.player_turn = 'O'
                        break
                    else:
                        print('The move is not valid! Try again.')

            else:
                (m, px, py) = self.max_alpha_beta(-2, 2)
                self.current_state[px][py] = 'O'
                self.player_turn = 'X'

In [ ]:
#Game 1
def main():
    g = Game()
    g.play_alpha_beta()

if __name__ == "__main__":
    main()

.| .| .| 
.| .| .| 
.| .| .| 

Evaluation time: 0.1077805s
Recommended move: X = 0, Y = 0
Insert the X coordinate: 0
Insert the Y coordinate: 0
X| .| .| 
.| .| .| 
.| .| .| 

X| .| .| 
.| O| .| 
.| .| .| 

Evaluation time: 0.0034053s
Recommended move: X = 0, Y = 1
Insert the X coordinate: 2
Insert the Y coordinate: 2
X| .| .| 
.| O| .| 
.| .| X| 

X| O| .| 
.| O| .| 
.| .| X| 

Evaluation time: 0.0005565s
Recommended move: X = 2, Y = 1
Insert the X coordinate: 2
Insert the Y coordinate: 1
X| O| .| 
.| O| .| 
.| X| X| 

X| O| .| 
.| O| .| 
O| X| X| 

Evaluation time: 5.72e-05s
Recommended move: X = 0, Y = 2
Insert the X coordinate: 0
Insert the Y coordinate: 2
X| O| X| 
.| O| .| 
O| X| X| 

X| O| X| 
.| O| O| 
O| X| X| 

Evaluation time: 1.67e-05s
Recommended move: X = 1, Y = 0
Insert the X coordinate: 1
Insert the Y coordinate: 0
X| O| X| 
X| O| O| 
O| X| X| 

It's a tie!
The total Evaluation time:0.1118162s


In [ ]:
#Game 2
def main():
    g = Game()
    g.play_alpha_beta()

if __name__ == "__main__":
    main()

.| .| .| 
.| .| .| 
.| .| .| 

Evaluation time: 0.0671s
Recommended move: X = 0, Y = 0
Insert the X coordinate: 1
Insert the Y coordinate: 1
.| .| .| 
.| X| .| 
.| .| .| 

O| .| .| 
.| X| .| 
.| .| .| 

Evaluation time: 0.0038922s
Recommended move: X = 0, Y = 1
Insert the X coordinate: 0
Insert the Y coordinate: 1
O| X| .| 
.| X| .| 
.| .| .| 

O| X| .| 
.| X| .| 
.| O| .| 

Evaluation time: 0.0006204s
Recommended move: X = 1, Y = 0
Insert the X coordinate: 0
Insert the Y coordinate: 2
O| X| X| 
.| X| .| 
.| O| .| 

O| X| X| 
.| X| .| 
O| O| .| 

Evaluation time: 5.65e-05s
Recommended move: X = 1, Y = 0
Insert the X coordinate: 1
Insert the Y coordinate: 0
O| X| X| 
X| X| .| 
O| O| .| 

O| X| X| 
X| X| .| 
O| O| O| 

The winner is O!
The total Evaluation time:0.0716691s


In [ ]:
#Game 3
def main():
    g = Game()
    g.play_alpha_beta()

if __name__ == "__main__":
    main()

.| .| .| 
.| .| .| 
.| .| .| 

Evaluation time: 0.0654421s
Recommended move: X = 0, Y = 0
Insert the X coordinate: 0
Insert the Y coordinate: 0
X| .| .| 
.| .| .| 
.| .| .| 

X| .| .| 
.| O| .| 
.| .| .| 

Evaluation time: 0.00402s
Recommended move: X = 0, Y = 1
Insert the X coordinate: 0
Insert the Y coordinate: 1
X| X| .| 
.| O| .| 
.| .| .| 

X| X| O| 
.| O| .| 
.| .| .| 

Evaluation time: 0.0003512s
Recommended move: X = 2, Y = 0
Insert the X coordinate: 2
Insert the Y coordinate: 0
X| X| O| 
.| O| .| 
X| .| .| 

X| X| O| 
O| O| .| 
X| .| .| 

Evaluation time: 6.75e-05s
Recommended move: X = 1, Y = 2
Insert the X coordinate: 1
Insert the Y coordinate: 2
X| X| O| 
O| O| X| 
X| .| .| 

X| X| O| 
O| O| X| 
X| O| .| 

Evaluation time: 1.72e-05s
Recommended move: X = 2, Y = 2
Insert the X coordinate: 2
Insert the Y coordinate: 2
X| X| O| 
O| O| X| 
X| O| X| 

It's a tie!
The total Evaluation time:0.0698979s


In [ ]:
#Game 4
def main():
    g = Game()
    g.play_alpha_beta()

if __name__ == "__main__":
    main()

.| .| .| 
.| .| .| 
.| .| .| 

Evaluation time: 0.107986s
Recommended move: X = 0, Y = 0
Insert the X coordinate: 0
Insert the Y coordinate: 0
X| .| .| 
.| .| .| 
.| .| .| 

X| .| .| 
.| O| .| 
.| .| .| 

Evaluation time: 0.0025427s
Recommended move: X = 0, Y = 1
Insert the X coordinate: 0
Insert the Y coordinate: 1
X| X| .| 
.| O| .| 
.| .| .| 

X| X| O| 
.| O| .| 
.| .| .| 

Evaluation time: 0.0003474s
Recommended move: X = 2, Y = 0
Insert the X coordinate: 2
Insert the Y coordinate: 0
X| X| O| 
.| O| .| 
X| .| .| 

X| X| O| 
O| O| .| 
X| .| .| 

Evaluation time: 0.0001848s
Recommended move: X = 1, Y = 2
Insert the X coordinate: 1
Insert the Y coordinate: 2
X| X| O| 
O| O| X| 
X| .| .| 

X| X| O| 
O| O| X| 
X| O| .| 

Evaluation time: 1.6e-05s
Recommended move: X = 2, Y = 2
Insert the X coordinate: 2
Insert the Y coordinate: 2
X| X| O| 
O| O| X| 
X| O| X| 

It's a tie!
The total Evaluation time:0.1110768s


In [ ]:
#Game 5
def main():
    g = Game()
    g.play_alpha_beta()

if __name__ == "__main__":
    main()

.| .| .| 
.| .| .| 
.| .| .| 

Evaluation time: 0.0655396s
Recommended move: X = 0, Y = 0
Insert the X coordinate: 2
Insert the Y coordinate: 2
.| .| .| 
.| .| .| 
.| .| X| 

.| .| .| 
.| O| .| 
.| .| X| 

Evaluation time: 0.0050392s
Recommended move: X = 0, Y = 0
Insert the X coordinate: 0
Insert the Y coordinate: 0
X| .| .| 
.| O| .| 
.| .| X| 

X| O| .| 
.| O| .| 
.| .| X| 

Evaluation time: 0.0005648s
Recommended move: X = 2, Y = 1
Insert the X coordinate: 2
Insert the Y coordinate: 1
X| O| .| 
.| O| .| 
.| X| X| 

X| O| .| 
.| O| .| 
O| X| X| 

Evaluation time: 7.32e-05s
Recommended move: X = 0, Y = 2
Insert the X coordinate: 0
Insert the Y coordinate: 2
X| O| X| 
.| O| .| 
O| X| X| 

X| O| X| 
.| O| O| 
O| X| X| 

Evaluation time: 1.55e-05s
Recommended move: X = 1, Y = 0
Insert the X coordinate: 1
Insert the Y coordinate: 0
X| O| X| 
X| O| O| 
O| X| X| 

It's a tie!
The total Evaluation time:0.0712323s


In [ ]:
#Game 6
def main():
    g = Game()
    g.play_alpha_beta()

if __name__ == "__main__":
    main()

.| .| .| 
.| .| .| 
.| .| .| 

Evaluation time: 0.0647323s
Recommended move: X = 0, Y = 0
Insert the X coordinate: 1
Insert the Y coordinate: 1
.| .| .| 
.| X| .| 
.| .| .| 

O| .| .| 
.| X| .| 
.| .| .| 

Evaluation time: 0.0039201s
Recommended move: X = 0, Y = 1
Insert the X coordinate: 0
Insert the Y coordinate: 1
O| X| .| 
.| X| .| 
.| .| .| 

O| X| .| 
.| X| .| 
.| O| .| 

Evaluation time: 0.0005965s
Recommended move: X = 1, Y = 0
Insert the X coordinate: 1
Insert the Y coordinate: 0
O| X| .| 
X| X| .| 
.| O| .| 

O| X| .| 
X| X| O| 
.| O| .| 

Evaluation time: 6.99e-05s
Recommended move: X = 0, Y = 2
Insert the X coordinate: 0
Insert the Y coordinate: 2
O| X| X| 
X| X| O| 
.| O| .| 

O| X| X| 
X| X| O| 
O| O| .| 

Evaluation time: 2.96e-05s
Recommended move: X = 2, Y = 2
Insert the X coordinate: 2
Insert the Y coordinate: 2
O| X| X| 
X| X| O| 
O| O| X| 

It's a tie!
The total Evaluation time:0.0693483s


In [ ]:
#Game 7
def main():
    g = Game()
    g.play_alpha_beta()

if __name__ == "__main__":
    main()

.| .| .| 
.| .| .| 
.| .| .| 

Evaluation time: 0.0682542s
Recommended move: X = 0, Y = 0
Insert the X coordinate: 2
Insert the Y coordinate: 0
.| .| .| 
.| .| .| 
X| .| .| 

.| .| .| 
.| O| .| 
X| .| .| 

Evaluation time: 0.009629s
Recommended move: X = 0, Y = 0
Insert the X coordinate: 0
Insert the Y coordinate: 0
X| .| .| 
.| O| .| 
X| .| .| 

X| .| .| 
O| O| .| 
X| .| .| 

Evaluation time: 0.0005291s
Recommended move: X = 1, Y = 2
Insert the X coordinate: 1
Insert the Y coordinate: 2
X| .| .| 
O| O| X| 
X| .| .| 

X| O| .| 
O| O| X| 
X| .| .| 

Evaluation time: 8.2e-05s
Recommended move: X = 2, Y = 1
Insert the X coordinate: 2
Insert the Y coordinate: 1
X| O| .| 
O| O| X| 
X| X| .| 

X| O| .| 
O| O| X| 
X| X| O| 

Evaluation time: 1.67e-05s
Recommended move: X = 0, Y = 2
Insert the X coordinate: 0
Insert the Y coordinate: 2
X| O| X| 
O| O| X| 
X| X| O| 

It's a tie!
The total Evaluation time:0.078511s


In [ ]:
#Game 8
def main():
    g = Game()
    g.play_alpha_beta()

if __name__ == "__main__":
    main()

.| .| .| 
.| .| .| 
.| .| .| 

Evaluation time: 0.0672743s
Recommended move: X = 0, Y = 0
Insert the X coordinate: 0
Insert the Y coordinate: 0
X| .| .| 
.| .| .| 
.| .| .| 

X| .| .| 
.| O| .| 
.| .| .| 

Evaluation time: 0.004276s
Recommended move: X = 0, Y = 1
Insert the X coordinate: 0
Insert the Y coordinate: 1
X| X| .| 
.| O| .| 
.| .| .| 

X| X| O| 
.| O| .| 
.| .| .| 

Evaluation time: 0.0002141s
Recommended move: X = 2, Y = 0
Insert the X coordinate: 2
Insert the Y coordinate: 0
X| X| O| 
.| O| .| 
X| .| .| 

X| X| O| 
O| O| .| 
X| .| .| 

Evaluation time: 6.56e-05s
Recommended move: X = 1, Y = 2
Insert the X coordinate: 1
Insert the Y coordinate: 2
X| X| O| 
O| O| X| 
X| .| .| 

X| X| O| 
O| O| X| 
X| O| .| 

Evaluation time: 1.79e-05s
Recommended move: X = 2, Y = 2
Insert the X coordinate: 2
Insert the Y coordinate: 2
X| X| O| 
O| O| X| 
X| O| X| 

It's a tie!
The total Evaluation time:0.0718479s


In [ ]:
#Game 9
def main():
    g = Game()
    g.play_alpha_beta()

if __name__ == "__main__":
    main()

.| .| .| 
.| .| .| 
.| .| .| 

Evaluation time: 0.0671561s
Recommended move: X = 0, Y = 0
Insert the X coordinate: 0
Insert the Y coordinate: 0
X| .| .| 
.| .| .| 
.| .| .| 

X| .| .| 
.| O| .| 
.| .| .| 

Evaluation time: 0.0026062s
Recommended move: X = 0, Y = 1
Insert the X coordinate: 0
Insert the Y coordinate: 1
X| X| .| 
.| O| .| 
.| .| .| 

X| X| O| 
.| O| .| 
.| .| .| 

Evaluation time: 0.0003793s
Recommended move: X = 2, Y = 0
Insert the X coordinate: 2
Insert the Y coordinate: 0
X| X| O| 
.| O| .| 
X| .| .| 

X| X| O| 
O| O| .| 
X| .| .| 

Evaluation time: 0.0001533s
Recommended move: X = 1, Y = 2
Insert the X coordinate: 1
Insert the Y coordinate: 2
X| X| O| 
O| O| X| 
X| .| .| 

X| X| O| 
O| O| X| 
X| O| .| 

Evaluation time: 1.62e-05s
Recommended move: X = 2, Y = 2
Insert the X coordinate: 2
Insert the Y coordinate: 2
X| X| O| 
O| O| X| 
X| O| X| 

It's a tie!
The total Evaluation time:0.0703111s


In [ ]:
#Game 10
def main():
    g = Game()
    g.play_alpha_beta()

if __name__ == "__main__":
    main()

.| .| .| 
.| .| .| 
.| .| .| 

Evaluation time: 0.1018848s
Recommended move: X = 0, Y = 0
Insert the X coordinate: 0
Insert the Y coordinate: 0
X| .| .| 
.| .| .| 
.| .| .| 

X| .| .| 
.| O| .| 
.| .| .| 

Evaluation time: 0.0027106s
Recommended move: X = 0, Y = 1
Insert the X coordinate: 0
Insert the Y coordinate: 1
X| X| .| 
.| O| .| 
.| .| .| 

X| X| O| 
.| O| .| 
.| .| .| 

Evaluation time: 0.000344s
Recommended move: X = 2, Y = 0
Insert the X coordinate: 2
Insert the Y coordinate: 0
X| X| O| 
.| O| .| 
X| .| .| 

X| X| O| 
O| O| .| 
X| .| .| 

Evaluation time: 6.96e-05s
Recommended move: X = 1, Y = 2
Insert the X coordinate: 1
Insert the Y coordinate: 2
X| X| O| 
O| O| X| 
X| .| .| 

X| X| O| 
O| O| X| 
X| O| .| 

Evaluation time: 1.65e-05s
Recommended move: X = 2, Y = 2
Insert the X coordinate: 2
Insert the Y coordinate: 2
X| X| O| 
O| O| X| 
X| O| X| 

It's a tie!
The total Evaluation time:0.1050255s


#Avg Comp Thik Time = (0.112 + 0.072 + 0.070 + 0.111 + 0.071 + 0.069 + 0.785 + 0.072 + 0.070 + 0.105)/ 10

#Avg Comp Thik Time = 0.218

#Alphabeta pruninng is faster than Minmax because based on the current state of the board it automatically deletes child nodes which is knows it will not take. Therefore, it visits less states and allows it to arrive to a logical answer to win the game faster. On the other hand, Minmax does not prune any of the branches that do not provide an optimal solution. Therefore, it has to spend more time calculating all the options. Therefore, it takes longer since it evaluates all possible states while Alphabeta, only considers the states after it removes the ones that it determines that will not give the agent an advantage